eth1_refactor/eth1client/eth1client.go

```diff
@@ -0,0 +1,268 @@
+	}
+
+	client := ec.client.Load()
+
+	ctx, cancel := context.WithTimeout(ctx, ec.connectionTimeout)
+	defer cancel()
+
+	sp, err := client.SyncProgress(ctx)
+	if err != nil {
+		ec.metrics.Eth1Failure()
+		return false, err
+	}
+
+	if sp != nil {
+		ec.metrics.Eth1Syncing()
+		return false, nil
+	}
+
+	ec.metrics.Eth1Ready()
+
+	return true, nil
+}
+
+func (ec *Eth1Client) isClosed() bool {
+	select {
+	case <-ec.closed:
+		return true
+	default:
+		return false
+	}
+}
+
+func (ec *Eth1Client) streamLogsToChan(ctx context.Context, logs chan ethtypes.Log) error {
+	client := ec.client.Load()
+
+	heads := make(chan *ethtypes.Header)
+
+	sub, err := client.SubscribeNewHead(ctx, heads)
+	if err != nil {
+		return fmt.Errorf("subscribe heads: %w", err)
+	}
+
+	for {
+		select {
+		case err := <-sub.Err():
+			return fmt.Errorf("subscription: %w", err)
+
+		case header := <-heads:
+			query := ethereum.FilterQuery{
+				Addresses: []ethcommon.Address{ec.contractAddress},
+				FromBlock: new(big.Int).SetUint64(ec.lastFetchedBlock.Load() + 1),
+				ToBlock:   header.Number,
+			}
+
+			newLogs, err := client.FilterLogs(ctx, query)
+			if err != nil {
+				return fmt.Errorf("fetch logs: %w", err)
+			}
+
+			for _, log := range newLogs {
+				logs <- log
+			}
+
+			ec.setLastFetchedBlock(query.ToBlock.Uint64())
+		}
+	}
+}
+
+// connect connects to eth1.
+// It must not be called twice in parallel.
+func (ec *Eth1Client) connect(ctx context.Context) error {
+	logger := ec.logger.With(fields.Address(ec.nodeAddr))
+
+	ctx, cancel := context.WithTimeout(ctx, ec.connectionTimeout)
+	defer cancel()
+
+	logger.Info("connecting", zap.Duration("timeout", ec.connectionTimeout))
+
+	client, err := ethclient.DialContext(ctx, ec.nodeAddr)
+	if err != nil {
+		logger.Error("connection failed", zap.Error(err))
+		return err
+	}
+
+	logger.Info("connected")
+	ec.client.Store(client)
+
+	return nil
+}
+
+// reconnect tries to reconnect multiple times with an exponent interval.
+// It panics when reconnecting limit is reached.
+// It must not be called twice in parallel.
+func (ec *Eth1Client) reconnect(ctx context.Context) {
+	logger := ec.logger.With(fields.Address(ec.nodeAddr))
+
+	if cl := ec.client.Load(); cl != nil {
+		cl.Close()
+	}
+
+	tasks.ExecWithInterval(func(lastTick time.Duration) (stop bool, cont bool) {
+		logger.Info("reconnecting")
+		if err := ec.connect(ctx); err != nil {
+			if ec.isClosed() {
+				return true, false
+			}
+			// continue until reaching to limit, and then panic as eth1 connection is required
+			if lastTick >= ec.reconnectionMaxInterval {
+				logger.Panic("failed to reconnect", zap.Error(err))
+			} else {
+				logger.Warn("could not reconnect, still trying", zap.Error(err))
+			}
+			return false, false
+		}
+		return true, false
+	}, ec.reconnectionInitialInterval, ec.reconnectionMaxInterval+(ec.reconnectionInitialInterval))
+
+	logger.Info("reconnected")
+}
+
+func (ec *Eth1Client) setLastFetchedBlock(block uint64) {
```

Comment:
I', thinking it might be better to store this in another component and keep Eth1 as just a client without logic.

Reply:
@y0sher so we extract `streamLogsToChan` to another component and keep the Eth1Client with `SubscribeNewHead` and `FilterLogs` methods instead?

Reply:
No, just pass the block number into the `SteamLogs` function. last fetched block should be saved and retrieved from the DB. (what used to be syncoffset). 
so the updating of it should be a part of the DB tx. 
Eth1 should be passed the required block to start from every time, it should not care about saving this.

Reply:
I see what you mean. Yes, we can actually extract it to a variable in `streamLogsToChan` and pass it to `StreamLogs`